### PR TITLE
ui: fee info and best unit functions

### PR DESCRIPTION
**Pull request by buck54321**
Adds a dialog to display fee information for a blockchain.

image

Adds `Doc` methods for selecting the most appropriate unit for a value. Adds a standardized element set that allows quickly setting values with units or rates with units. Adds hover menu to allow user to select different units for a displayed value. Populates alternative units in various `UnitInfo`.

### CODE REVIEW

File: client/asset/zec/zec.go [hunk 1]
```diff
@@ -320,7 +320,7 @@ func (w *zecWallet) CallRPC(method string, args []any, thing any) error {
 
 // FeeRate returns the asset standard fee rate for Zcash.
 func (w *zecWallet) FeeRate() uint64 {
-	return dexzec.LegacyFeeRate
+	return 5000 // per logical action
```
Comment (ukane-philemon) [inline client/asset/zec/zec.go:323]: Why is `dexzec.LegacyFeeRate` no longer suitable?

File: client/core/trade.go [hunk 2]
```diff
@@ -377,9 +377,9 @@ func (t *trackedTrade) cacheRedemptionFeeSuggestion() {
 	// to an external fee oracle if an internal estimate is not available and
 	// the wallet settings permit external API requests.
 	toWallet := t.wallets.toWallet
-	if rater, is := toWallet.feeRater(); is && t.readyToTick && toWallet.connected() {
-		if feeRate := rater.FeeRate(); feeRate != 0 {
-			set(feeRate)
+	if t.readyToTick && toWallet.connected() {
+		if r := toWallet.feeRate(); r != 0 {
```
Comment (ukane-philemon) [inline client/core/trade.go:381]: I'd prefer `feeRate` to `r`

File: client/core/wallet.go [hunk 3]
```diff
@@ -279,6 +281,10 @@ func (w *xcWallet) state() *WalletState {
 	if w.connector.On() {
 		tokenApprovals = w.ApprovalStatus()
 	}
+	var feeState *FeeState
```
Comment (ukane-philemon) [inline client/core/wallet.go:284]: You'd want a new line above this, I think.

**Review by ukane-philemon — COMMENTED** [review]
I'm running this PR but can't see the fee box. Do you know if there is a setting I need?

**Comment by buck54321** [comment]
> I'm running this PR but can't see the fee box. Do you know if there is a setting I need?

It's only displayed if you have both fiat exchange rate and a fee rate **from the wallet**. If you're on an SPV wallet, you must have external rates allowed (probably all utxo wallets on simnet, actually). You also must have fiat exchange rates on.

**Comment by ukane-philemon** [comment]
> If you're on an SPV wallet, you must have external rates allowed (probably all utxo wallets on simnet, actually).

I'm running testnet and all these settings are on 🥲  but it's still not showing up.

File: client/asset/zec/zec.go (hunk 1 above)
Comment (buck54321) [reply client/asset/zec/zec.go:323]: That's not how fees work for Zcash anymore. See [ZIP-0317](https://zips.z.cash/zip-0317).

**Comment by buck54321** [comment]
I'll hit you on Matrix

**Comment by ukane-philemon** [comment]
> I'll hit you on Matrix

Thanks! Works now, I wasn't running the correct exec.

I see this weird behaviour. The popup refused to close and I could open more and more on top of the previous popup.
Screenshot 2024-05-06 at 12 28 36 AM

**Comment by ukane-philemon** [comment]
`feeState` has always been `null` for `btc`. Do I need to author a tx before it shows up? I can only see `Unable to get optimal fee rate, using fallback of 4539978304` log.

Testing well for `dcr`.

**Review by ukane-philemon — APPROVED** [review]
Looking good, but still not able to view the fee box for `btc` and `bch`.

**Comment by JoeGruffins** [comment]
I see the fee boxes. @ukane-philemon are you sure your site files are on the right commit? There's a Build ID at the bottom of the Settings menu you can check to make sure:

image

**Comment by ukane-philemon** [comment]
I had to toggle the external fee option, reconfigure and restart before it showed up for btc.

```
2024-05-07 11:17:40.506 [DBG] CORE[dcr]: tip change: 1382877 (ce80dfca666f121f63b6d246c4505f4ed66909ad2a069055caca007ea53fa380) => 1382878 (e283aa45893b0d72bcca3f55ca2033ea08812b9c96e25186efdc291f36641170)
2024-05-07 11:17:40.528 [DBG] CORE[dcr][SPV]: Starting cspp funds mixer with mix.decred.org:15760
2024-05-07 11:18:46.752 [INF] CORE: notify: |SUCCESS| (walletconfig) Wallet configuration updated - Configuration for btc wallet has been updated. Deposit address = tb1qvztsnndk2tghkueedvnvxv8c4kenm8kxu2g9m7
2024-05-07 11:18:46.752 [INF] CORE: btc wallet configuration updated without a restart 👍

```

Screenshot 2024-05-07 at 12 21 54 PM

Screenshot 2024-05-07 at 12 22 14 PM

**Comment by ukane-philemon** [comment]
I'm yet to reconfigure BCH, It's still in the previous state as BTC. For some reasons, it seems the `apifeefallback` is not properly populated until a reconfigure. 

Note: I haven't used this testnet wallet for months. What changed?

Screenshot 2024-05-07 at 12 24 17 PM

File: dex/networks/eth/params.go [hunk 4]
```diff
@@ -147,29 +158,47 @@ func RefundGas(contractVer uint32) uint64 {
...
+// ceiling.
+func WeiToGweiSafe(wei *big.Int) (uint64, error) {
 	if wei.Cmp(new(big.Int)) == -1 {
 		return 0, fmt.Errorf("wei must be non-negative")
 	}
-	gweiFactorBig := big.NewInt(GweiFactor)
-	gwei := new(big.Int).Div(wei, gweiFactorBig)
+	gwei := new(big.Int).Div(wei.Add(wei, gweiCeilAddend), gweiFactorBig)
 	if !gwei.IsUint64() {
 		return 0, fmt.Errorf("suggest gas price %v gwei is too big for a uint64", wei)
```
Comment (JoeGruffins) [inline dex/networks/eth/params.go:203]: nit:
```suggestion
		return 0, fmt.Errorf("%v gwei is too big for a uint64", gwei)
```

File: dex/networks/eth/params.go [hunk 5]
```diff
@@ -147,29 +158,47 @@ func RefundGas(contractVer uint32) uint64 {
...
-func WeiToGweiUint64(wei *big.Int) (uint64, error) {
+// For values that are not even multiples of 1 gwei, this function returns the
+// ceiling.
+func WeiToGweiSafe(wei *big.Int) (uint64, error) {
 	if wei.Cmp(new(big.Int)) == -1 {
 		return 0, fmt.Errorf("wei must be non-negative")
 	}
-	gweiFactorBig := big.NewInt(GweiFactor)
-	gwei := new(big.Int).Div(wei, gweiFactorBig)
+	gwei := new(big.Int).Div(wei.Add(wei, gweiCeilAddend), gweiFactorBig)
```
Comment (JoeGruffins) [inline dex/networks/eth/params.go:201]: I think this alters the passed wei value if that matters. https://go.dev/play/p/yonPWn8AAEt

Same for the new WeiToGweiCeil and the passed v

Could add a comment about it.

**Review by JoeGruffins — APPROVED** [review]

File: dex/networks/eth/params.go (hunk 5 above)
Comment (buck54321) [reply dex/networks/eth/params.go:201]: Yes it does. I think it's righteous, but I'll do an audit and leave comments.

**Comment by ukane-philemon** [comment]
@buck54321 can we use `Errorf` for the log in  `func (btc *baseWallet) FeeRate() uint64`? Maybe you can add to your PR in your next commit.

File: client/asset/interface.go [hunk 6]
```diff
@@ -560,6 +564,8 @@ type Wallet interface {
 	SingleLotSwapRefundFees(version uint32, feeRate uint64, useSafeTxSize bool) (uint64, uint64, error)
 	// SingleLotRedeemFees returns the fees for a redeem transaction for a single lot.
 	SingleLotRedeemFees(version uint32, feeRate uint64) (uint64, error)
+	// StandardSendFee returns the send fee for a "standard" tx fee.
```
Comment (martonp) [inline client/asset/interface.go:567]:
```suggestion
	// StandardSendFee returns the fee for a "standard" send tx.
```

File: client/asset/eth/eth.go [hunk 7]
```diff
@@ -3090,12 +3120,16 @@ func (w *TokenWallet) EstimateSendTxFee(addr string, value, _ uint64, subtract b
...
+	maxFee, _, _, err := w.canSend(value, addr != "", true)
 	if err != nil {
 		return 0, false, err
 	}
 	return maxFee, w.ValidateAddress(addr), nil
+}
 
+// StandardSendFees returns the fees for a simple send tx.
+func (w *TokenWallet) StandardSendFee(feeRate uint64) uint64 {
+	return defaultSendGasLimit * feeRate
```
Comment (martonp) [inline client/asset/eth/eth.go:3132]: You need to use the token send gas instead of `defaultSendGasLimit`.

**Review by martonp — COMMENTED** [review]
Looks good, I just don't see the fees displayed for DCR. Maybe a UI issue since DCR has additional panels there?

**Comment by buck54321** [comment]
> func (btc *baseWallet) FeeRate() uint64

We can't do it at an error because it's not necessarily an error. For an spv wallet or an rpc wallet without a primed `estimatesmartfee`, the inability to get a fee rate ourselves is ok.

**Review by martonp — APPROVED** [review]

**Pull request merged**